### ISSUE TITLE
Prefix added to version of packages in package.json

### ISSUE THREAD

**Issue by rockchalkwushock** [yarnpkg/yarn#3787]
```plaintext
yarn@0.27.5
node@8.1.3
npm @5.0.3
macOS@10.12.5

yarn bin: /Users/codybrunner/node_modules/.bin
yarn global bin: /Users/codybrunner/.node_modules_global/bin
```

I did not see a similar issue opened and this could just be based on my own local settings with `yarn`. Seems after upgrading to `@0.27.5` whenever I install a package I get the following:

```js
{
  dependencies: {
    react: "/Users/codybrunner15.16.1" // expect react: "~15.16.1"
  }
}
```

My guess is that `yarn` is now using the whole prefix of where to find the cached modules on my machine. Not very well versed in my understanding of how some of that works 😕 . Figured I'd report this...if it is my own settings causing this feel free to close.

**Comment by rockchalkwushock** [comment]
I just noticed that support was added in `@0.27.0` for `~` to be seen as `$HOME`. I'm now wondering if the fact I have `save-prefix: "~"` could be resulting in a translation from `~15.16.1` to `/Users/codybrunner15.16.1`. I'm going to check this out and get back with an answer soon 😉

**Comment by rockchalkwushock** [comment]
Yep absolutely the problem!

After running:

```sh
yarn config set save-prefix "^"
mkdir test
cd test
yarn init
yarn add react
```

## Result

```js
{
  "dependencies": {
    "react": "^15.16.1" // expected result!
  }
}
```

Unfortunately I have not the slightest idea of how to resolve something like this in the underlying source code :confused:; but at least I now know that the change made in `@0.270` to support `~` being seen as `$HOME` is the underlying issue. I will leave this to someone on the `yarn` team or someone that knows more about how and where to fix this in the source code...if it's something that should be fixed.

**Comment by arcanis** [comment]
I guess the easy fix would be to check for `~/` instead of `~`. Thanks for investigating!

**Comment by rockchalkwushock** [comment]
@arcanis 

I figured it would be something like this. I've never done much with CLI related libraries so I was hesitant to make any changes. I'll open a PR if I can find where in the source code to make the change and submit when the tests are passing locally.

**Comment by rockchalkwushock** [comment]
DISCLAIMER: I am not strong in my regex-fu:

```js
path = path.replace(/^\s*~(?=$|\/|\\)/, userHome);
```

Not understanding what the initial intent here was 😕. If `userHome` is whatever the users home directory is whether `linux` or `macOS` Then I'm reading this as we should replace the provided string argument with the user's home directory matched to `~ || ~/`. What I'm not understanding is that after some reading `(?=ABC)` is a positive look ahead that:

> Matches a group after the main expression without including it in the result.

If we are not including it in the result then how is the OR logic even being applied? I could just be way off base with reading this as I said the regex-fu is weak with this one.

To me it seems that `path` will always be equal to `userHome` or the string `/Users/<user>`.

Any pointers, advice, etc as I attempt to wrap my head around this code?

**Comment by cwp** [comment]
This also affects `yarn install`. If you have a dependency in package.json that begins with `~` yarn will replace the `~` with `userHome`, and then complain about the invalid version. Something like:

```
Couldn't find any versions for "bluebird" that matches "/Users/colin3.5.0"
```

**Comment by rockchalkwushock** [comment]
I imagine the same holds true for:

```sh
yarn add --tilde <package>
```

**Comment by cwp** [comment]
(Deleted report of `yarn install` being affected. It was just corrupted package.json files in the cache.)

**Comment by ryardley** [comment]
This looks like where it happened: https://github.com/yarnpkg/yarn/commit/35ee7de206e3078df32f1626d1dde6ba1ed41d0b

It seems there is an assumption that all values within the config are paths. Probably want to set out a whitelist or have a config schema things are built off so only path type values are normalized.

**Comment by rockchalkwushock** [comment]
@ryardley 

Yeah I believe it stems from #630 this [comment](https://github.com/yarnpkg/yarn/issues/630#issuecomment-277837483).

**Comment by ryardley** [comment]
@rockchalkwushock yes I just noticed it is older than that one above trying to work out which version to install.

**Comment by danoc** [comment]
Looks like this was fixed in 1.0 and can be closed? cc: @BYK

**Comment by BYK** [comment]
Yup. Thanks for the heads up!
